src/engraving/io/msczreader.cpp

```diff
@@ -0,0 +1,189 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-CLA-applies
+ *
+ * MuseScore
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2021 MuseScore BVBA and others
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include "msczreader.h"
+
+#include <QXmlStreamReader>
+#include <QFile>
+#include <QFileInfo>
+
+#include "thirdparty/qzip/qzipreader_p.h"
+
+#include "log.h"
+
+//! NOTE The current implementation resolves files by extension.
+//! This will probably be changed in the future.
+
+using namespace mu::engraving;
+
+MsczReader::MsczReader(const QString& filePath)
+    : m_filePath(filePath), m_device(new QFile(filePath)), m_selfDeviceOwner(true)

```

Comment:
Are we still using raw pointers for new code then?  Would seem to be some opportunities for using unique_ptr etc. here.

Reply:
`m_device` can be created here or passed from outside (another constructor or setter). We can require a shared point, but in practice, it is not convenient in this case. Because a typical use case is as follows:
```
QByteArray data = ...
QBuffer buf(&data);
MsczReader reader(&buf);
```

Reply:
Yes I understand that, but in the case it does need to create/manage its own device, was thinking unique_ptr or similar would be appropriate.